server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,242 @@
+            .features(Scalar.Feature.DETERMINISTIC)
+            .typeVariableConstraints(typeVariable("V"))
+            .build();
+
+    static final Signature PARAMETER_SIGNATURE =
+        Signature.builder(NAME, FunctionType.AGGREGATE)
+            .argumentTypes(TypeSignature.parse("V"),
+                DataTypes.INTEGER.getTypeSignature())
+            .returnType(new ArrayType<>(DataTypes.UNTYPED_OBJECT).getTypeSignature())
+            .features(Scalar.Feature.DETERMINISTIC)
+            .typeVariableConstraints(typeVariable("V"))
+            .build();
+
+    static {
+        DataTypes.register(TopKAggregation.TopKStateType.ID, in -> TopKAggregation.TopKStateType.INSTANCE);
+    }
+
+    public static void register(Functions.Builder builder) {
+        builder.add(
+            DEFAULT_SIGNATURE,
+            TopKAggregation::new
+        );
+
+        builder.add(
+            PARAMETER_SIGNATURE,
+            TopKAggregation::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+    private static final int DEFAULT_MAP_SIZE = 32;
+    private static final int DEFAULT_LIMIT = 8;
+
+    private TopKAggregation(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+
+    @Nullable
+    @Override
+    public TopKState newState(RamAccounting ramAccounting,
+                              Version indexVersionCreated,
+                              Version minNodeInCluster,
+                              MemoryManager memoryManager) {
+        ramAccounting.addBytes(calculateRamUsage(DEFAULT_MAP_SIZE));
+        return new TopKState(new ItemsSketch<>(DEFAULT_MAP_SIZE), DEFAULT_LIMIT);
+    }
+
+    @Override
+    public TopKState iterate(RamAccounting ramAccounting,
+                             MemoryManager memoryManager,
+                             TopKState state,
+                             Input<?>... args) throws CircuitBreakingException {
+        if (state.itemsSketch.isEmpty() && args.length == 2) {
+            Integer limit = (Integer) args[1].value();
+            if (limit <= 0) {
+                throw new IllegalArgumentException("Invalid value for parameter k for top-k aggregation");
+            }
+            int maxMapSize = maxMapSize(limit);
+            // Calculate new max space usage and adapt ram accounting
+            long newRamUsage = calculateRamUsage(maxMapSize);
+            long defaultRamUsage = calculateRamUsage(DEFAULT_MAP_SIZE);
+            long diff = newRamUsage > defaultRamUsage ? newRamUsage - defaultRamUsage : defaultRamUsage - newRamUsage;
+            ramAccounting.addBytes(diff);
```

Comment:
If the new one is smaller, we must remove bytes. This can be simply done by adding a negative number.
```suggestion
            ramAccounting.addBytes(newRamUsage - defaultRamUsage);
```